pages/react-canvas-components.mdx

```diff
@@ -0,0 +1,256 @@
+
+context.beginPath();
+corners.forEach((corner, index) => {
+  if (index === 0) {
+    context.moveTo(corner.x, corner.y);
+  } else {
+    context.lineTo(corner.x, corner.y);
+  }
+});
+context.fillStyle = color;
+context.fill();
+```
+
+Now, how could we extract this logic into its own `Hexagon` component ? We need the canvas' context to draw anything on it. We could pass it to all the child components via a prop, but this approach can be a bit tedious when children are nested. Another way of doing this is by using a [React context](https://reactjs.org/docs/context.html) to share "global" values between components.
+
+## 📦 A context in a context
+
+At this point the naming gets a bit tricky as we are trying to share a canvas' _context_ via a React _context_. Once we create a React context, we need to use the context's `Provider` to share a value. In the case of our `Canvas` component, which would look like this:
+
+```jsx hexagons-30dkf src/Canvas.js
+// we create a React context with a _null_ default value
+const SharingContext = React.createContext(null);
+
+const Canvas = props => {
+  const canvasRef = React.useRef(null);
+  const [renderingContext, setRenderingContext] = React.useState(null);
+
+  // the canvas rendering context is not immediately avalaible
+  // as the canvas node first needs to be added to the DOM by react
+  React.useEffect(() => {
+    const context2d = canvasRef.current.getContext('2d');
+    setRenderingContext(context2d);
+  }, []);
+
+  return (
+    <SharingContext.Provider value={renderingContext}>
+      <canvas ref={canvasRef} />
+      {/* the hexagons are passed through the `children` prop */}
+      {props.children}
+    </SharingContext.Provider>
+  );
+};
+```
+
+The `Hexagon` component needs to consume this React context to read its value – here with the `useContext` hook.
+
+```jsx hexagons-30dkf src/Hexagon.js
+const Hexagon = props => {
+  // we get the rendering context by comsuming the React context
+  const renderingContext = React.useContext(SharingContext);
+
+  if (renderingContext !== null) {
+    // hexagon drawing logic
+  }
+};
+```
+
+Now, both our `Canvas` and `Hexagon` components are ready. We can display several random hexagons:
+
+```jsx hexagons-30dkf src/App.js
+const App = () => (
+  <Canvas>
+    {getRandomHexagons().map(hexagon => (
+      <Hexagon {...hexagon} />
+    ))}
+  </Canvas>
+);
+```
+
+The last thing we need is to animate the hexagons so that they rotate.
+
+## 🎬 Animations
+
+As we saw, the canvas is like a sheet of paper: once we've drawn on it, we can't change it! But we can clear the canvas and re-draw on it. Animating on a canvas is like animating cartoons. It's a lots of draw, clean, draw, clean, repeat! To make a shape move, you need to split the movement into small steps, draw them one by one, while clearing the canvas between them. Those steps are called frames. Browsers come with an API `requestAnimationFrame` so that you can draw in each frame. 
```

Comment:
- I'd use a dash here instead of a colon
- "once it's been drawn on, it can't be changed" 
- "However, a canvas can be cleared in order that something new can be draw on it. In that respect animating a canvas is somewhat like old-fashioned cartoon animation - we draw, clean, draw, clean and repeat until we achieve the desired effect" 
- " while clearing the canvas in-between"